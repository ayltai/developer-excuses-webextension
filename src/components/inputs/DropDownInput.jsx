import { Box, FormControl, InputLabel, makeStyles, MenuItem, Select, } from '@material-ui/core';
import PropTypes from 'prop-types';
import React from 'react';

import { uuid, } from '../../Utils';

export const DropDownInput = props => {
    const classes = makeStyles(theme => ({
        margin    : {
            marginBottom : theme.spacing(2),
        },
        fullWidth : {
            width : '100%',
        },
    }))();

    const [ value, setValue, ] = React.useState(props.value);

    const id = React.useMemo(() => uuid(), []);

    return (
        <Box className={classes.margin}>
            <FormControl
                className={classes.fullWidth}
                variant='filled'>
                <InputLabel id={id}>
                    {props.label}
                </InputLabel>
                <Select
                    labelId={id}
                    value={value}
                    onChange={event => {
                        setValue(event.target.value);

                        if (props.onChange) props.onChange(event.target.value);
                    }}>
                    {props.entries.map(entry => (
                        <MenuItem
                            key={entry.value}
                            value={entry.value}>
                            {entry.label}
                        </MenuItem>
                    ))}
                </Select>
            </FormControl>
        </Box>
    );
};

DropDownInput.propTypes = {
    label    : PropTypes.string,
    entries  : PropTypes.arrayOf(PropTypes.object).isRequired,
    value    : PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]),
    onChange : PropTypes.func,
};
