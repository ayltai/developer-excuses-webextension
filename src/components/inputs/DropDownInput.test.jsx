import { FormControl, Select } from '@material-ui/core';
import { mount, shallow, } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import { DropDownInput, } from './DropDownInput';

jest.mock('../../Utils');

describe('<DropDownInput />', () => {
    it('renders correctly', () => {
        expect(renderer.create(
            <DropDownInput
                label='dummy'
                entries={[
                    {
                        label : '1',
                        value : 'a',
                    },
                    {
                        label : '2',
                        value : 'b',
                    }
                ]}
                value='a'
                onChange={() => {}} />
        ).toJSON()).toMatchSnapshot();
    });

    it('mounts without error', () => {
        mount(
            <DropDownInput
                label='dummy'
                entries={[
                    {
                        label : '1',
                        value : 'a',
                    },
                    {
                        label : '2',
                        value : 'b',
                    }
                ]}
                value='a'
                onChange={() => {}} />
        );
    });

    it('triggers onChange when changed', () => {
        const onChange = jest.fn();

        shallow(
            <DropDownInput
                label='dummy'
                entries={[
                    {
                        label : '1',
                        value : 'a',
                    },
                    {
                        label : '2',
                        value : 'b',
                    }
                ]}
                value='a'
                onChange={onChange} />)
            .find(FormControl)
            .find(Select)
            .simulate('change', {
                target : {
                    value : 'b',
                },
            });

        expect(onChange.mock.calls.length).toEqual(1);
    });

    it('triggers onChange when changed', () => {
        let value = 'a';

        const onChange = newValue => value = newValue;

        shallow(
            <DropDownInput
                label='dummy'
                entries={[
                    {
                        label : '1',
                        value : 'a',
                    },
                    {
                        label : '2',
                        value : 'b',
                    }
                ]}
                value='a'
                onChange={onChange} />)
            .find(FormControl)
            .find(Select)
            .simulate('change', {
                target : {
                    value : 'b',
                },
            });

        expect(value).toEqual('b');
    });

    it('does not trigger onChange when changed', () => {
        let value = 'a';

        shallow(
            <DropDownInput
                label='dummy'
                entries={[
                    {
                        label : '1',
                        value : 'a',
                    },
                    {
                        label : '2',
                        value : 'b',
                    }
                ]}
                value='a' />)
            .find(FormControl)
            .find(Select)
            .simulate('change', {
                target : {
                    value : 'b',
                },
            });

        expect(value).toEqual('a');
    });
});
