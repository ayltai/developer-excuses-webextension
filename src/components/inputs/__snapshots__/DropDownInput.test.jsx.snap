// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropDownInput /> renders correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-4 makeStyles-margin-1"
>
  <div
    className="MuiFormControl-root makeStyles-fullWidth-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
      data-shrink={true}
      id="dummy"
    >
      dummy
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="dummy"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        1
      </div>
      <input
        onAnimationStart={[Function]}
        type="hidden"
        value="a"
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
